sign/dilithium/mode3/internal/sample.go

```diff
@@ -2,9 +2,192 @@ package internal
 
 import (
 	"github.com/cloudflare/circl/internal/shake"
-	common "github.com/cloudflare/circl/sign/dilithium/internal"
+	"github.com/cloudflare/circl/sign/dilithium/internal/f1600x4"
+
+	"github.com/cloudflare/circl/sign/dilithium/internal/common"
+
+	"encoding/binary"
 )
 
+// DeriveX4Available indicates whether the system supports the quick fourway
+// sampling variants like PolyDeriveUniformX4.
+var DeriveX4Available = f1600x4.Available && !UseAES
+
+// For each i, sample ps[i] uniformly with coefficients of norm less than γ₁
+// using the the given seed and nonces[i].  ps[i] may be nil and is ignored
+// in that case.  ps[i] will not be normalized, but have coefficients in the
+// interval (q-γ₁,q+γ₁).
+//
+// Can only be called when DeriveX4Available is true.
+func PolyDeriveUniformLeGamma1X4(ps [4]*common.Poly, seed *[48]byte,
+	nonces [4]uint16) {
+	var perm f1600x4.State
+	state := perm.Initialize()
+
+	// Absorb the seed in the four states
+	for i := 0; i < 6; i++ {
+		v := binary.LittleEndian.Uint64(seed[8*i : 8*(i+1)])
+		for j := 0; j < 4; j++ {
+			state[i*4+j] = v
+		}
+	}
+
+	// Absorb the nonces, the SHAKE256 domain separator (0b1111), the
+	// start of the padding (0b...001) and the end of the padding 0b100...
+	// Recall that the rate of SHAKE256 is 136 --- i.e. 17 uint64s.
+	for j := 0; j < 4; j++ {
+		state[6*4+j] = uint64(nonces[j]) | (0x1f << 16)
+		state[16*4+j] = 0x80 << 56
+	}
+
+	var idx [4]int // indices into ps
+	for j := 0; j < 4; j++ {
+		if ps[j] == nil {
+			idx[j] = 256 // mark nil polynomial as completed
+		}
+	}
+
+	// Each try requires 15 bits.  15 does not divide into 64, nor in 136,
+	// so we will have to carry some bits from a previous uint64 to the next.
+	var carry [4]uint32
+
+	// Shift is the amount of bits in the carry.
+	var shift [4]uint
+
+	for {
```

Comment:
Make this loop be more easily to read. 
Also beware that `break` in Go sometimes requires to specify a label.

```suggestion
	for !done { 
```

Reply:
Thanks, I improved it.